pkg/controller/nstemplateset/cluster_resources.go

```diff
@@ -2,82 +2,160 @@ package nstemplateset
 
 import (
 	"context"
+	"fmt"
 
 	toolchainv1alpha1 "github.com/codeready-toolchain/api/pkg/apis/toolchain/v1alpha1"
 	applycl "github.com/codeready-toolchain/toolchain-common/pkg/client"
+	"github.com/codeready-toolchain/toolchain-common/pkg/template"
 	"github.com/go-logr/logr"
 	quotav1 "github.com/openshift/api/quota/v1"
+	errs "github.com/pkg/errors"
+	"github.com/redhat-cop/operator-utils/pkg/util"
+	rbacv1 "k8s.io/api/rbac/v1"
 	"k8s.io/apimachinery/pkg/api/errors"
-	"k8s.io/apimachinery/pkg/api/meta"
 	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
 	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
 )
 
 type clusterResourcesManager struct {
 	*statusManager
 }
 
-// ensure ensures that the cluster resources exists.
+// listExistingResources returns a list of comparable  ToolchainObjects representing existing resources in the cluster
+type listExistingResources func(cl client.Client, username string) ([]applycl.ComparableToolchainObject, error)
+
+// toolchainObjectKind represents a resource kind that should be present in templates containing cluster resources.
+// Such a kind should be watched by NSTempalateSet controller which means that every change of the
+// resources of that kind triggers a new reconcile.
+// It is expected that the template contains only the kinds that are being watched and there is an instance of
+// toolchainObjectKind type created in clusterResourceKinds list
+type toolchainObjectKind struct {
+	gvk                   schema.GroupVersionKind
+	object                runtime.Object
```

Comment:
I kept the original `object` because I didn't combine the type with other types/objects, so I thought that there shouldn't be any confusion.
Anyway, I renamed it in: 317e5b0